### DISCUSSION TITLE
Wanted: more parameters for custom commands

### DISCUSSION THREAD

**Discussion by vladstudio** [notaapp/talk#266] [Feature Requests]
So I got very excited when I learned that I could extend Nota by writing custom commands... until I found that the only data available is "file" and "workspace". Don't get me wrong, great start! :-)

Here is what I'd looove to be able to access and modify:

- "**selections**". Not unlike "selectedText" in paste hook, but: 
    - an array - since Nota supports multiple selections;
    - editable - so that I can modify contents of each selection.
- "**cursors**" (or "**carets**"):
    - again, an array;
    - function to add content before/after a cursor (as requested @ https://github.com/notaapp/talk/discussions/205 ).

**Comment by vladstudio** [comment]
P.S. I do realize that this is not a trivial request - as I understand, right now data only flows one way, a script gets data but cannot send anything back.

Here is a quick idea how it might work:

Imagine I have thw following markdown file called "test.md":

```
test one
test two
test three
```

And I also have a custom command called "Transform to Uppercase". 
I open "test.md" in Nota, select second line ("test two"), open a command palette and run "Transform to Uppercase".

I imagine the script could receive something like this:

file='/path/to/test.md'
selections[0][start]=9 (that's the offset of first selection start)
selections[0][length]=8
selections[0][text]="test two" (redundant but convenient)

Then it would be my job to update symbols 9 to 17 with "TEST TWO" (with js or bash).

It is an ugly hack though :-) And for it to work properly, I'd need to process selections from last to first.

What do you think?

**Reply by vladstudio** [reply]
year Uppercase is bad example, cause Nota already can do it :-)

**Comment by astoilkov** [comment]
I like these ideas a lot. I want this kind of extensibility as well. However, I'm not sure how to handle arrays in scripts. It looks very complicated to write such a script and defeats the purpose of script extensibility simplicity.

What I'm thinking is that we could enable this kind of extensibility through a JavaScript API. Something like:
```js
export default function toUpperCase({ selections }) {
  selections[0].start
  return {
     type: 'replace-text',
     start,
     end,
     text
  }
}
```

**Reply by vladstudio** [reply]
I totally agree. Being able to work with selections and cursors is great, but it makes things much more complex. 

Sorry I'm not sure I understood your code sample! But in turn, let me fantasize - what if Nota could expose the following Javascript objects and functions to custom commands:

```js
Nota.activeFile.title; // "Test title", maybe editable
Nota.activeFile.path; // "/_Inbox/test.md", maybe editable, or:
Nota.activeFile.move(new_path);
Nota.activeFile.delete();

Nota.activeFile.selections; // array
Nota.activeFile.selections[0].text; // string, maybe editable

Nota.activeFile.cursors; // array
Nota.activeFile.cursors[0].insert(str); // after cursor
Nota.activeFile.cursors[0].delete(number); // before cursor
```

Then, for example, the custom command to Insert Sequential Numbers on all cursors would look like:

```
let c = Nota.activeFile.cursors;
for (var i = c.length - 1; i >= 0; i--) {
    c.insert(i+1);
}
```

P.S. I have no idea how complex it is to implement, please use scepticism.

P.P.S. Some API inspiration: https://www.sceditor.com/api/sceditor/

**Reply by borekb** [reply]
@astoilkov I think having a JavaScript API would be great 👍.

**Reply by astoilkov** [reply]
Yes, your API also works. We will probably start exploring this type of extensibility in the next few months.
